### DISCUSSION TITLE
Incremental port from WinForms to Avalonia

### DISCUSSION THREAD

**Discussion by wsandy1** [AvaloniaUI/Avalonia#11104] [Q&A] [answered]
Hey there,

I'm working as a developer on a long running project which uses incredibly messy WinForms display code mixed in with application code. We're looking into porting to Avalonia to allow us to A: separate the application and views, B: allow us to release a mobile version of our application. 

Because of the state of the program porting over all at once will be a big job - so is it possible to port from WinForms to avalonia in an incremental way?

Thanks:)

**Comment by robloo** [comment, marked as answer]
It's possible to embed Avalonia within WinForms similar to embedding WPF in WinForms. See this control: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32.Interop/WinForms/WinFormsAvaloniaControlHost.cs

> long running project which uses incredibly messy WinForms display code mixed in with application code. 

Perfectly describes every WinForms LOB application I've seen.

> A: separate the application and views,

I've ported WinForms applications from spaghetti code to WPF MVVM before. You will just have to go form-by-form converting it as it makes sense to models, views and view models. There is no easy shortcut here and it almost always results in a near 100% rewrite to do it correctly. In fact, it would be faster/better to rewrite except for the potential loss of esoteric business logic.

> B: allow us to release a mobile version of our application.

Unfortunately, it's all or nothing until you complete the conversion. You can modernize piece-by-piece but until it is all ported over to Avalonia you aren't going to get any mobile platforms working. The good news is for existing Windows desktop it can continue to work during porting.

**Reply by wsandy1** [reply]
Thanks, this is super helpful. It’s not a problem for us to not have a working mobile app until the end - it’s just easier for us to develop one codebase and not have to play “catch up” on the new codebase as new features are added to the old codebase. Thanks!

**Reply by robloo** [reply]
Glad it helps. Also keep in mind it's best to port as follows. (These are some past lessons learned that might help)

 1. Try to extract as much business logic and models cleanly from the existing code base. Refactor the existing code to do this.
 2. If your application is separated into several "pages" as most are. Convert each page to MVVM (along with all its children) separately. Then only the parent page is hosted by the WinForms control. Generally you don't want to host individual Avalonia controls in WinForms, you want to host the entire views.
 3. Find any self contained child views and do those first. These are typically separate windows or dialogs but may also be custom controls.
 4. points two and three above are a bit recursive and application specific. Just think about where it makes sense to break apart your application so you can do it piece by piece with children done first.
 5. Do the view models right this time and you won't ever need to do this level of refactoring and rewrites again. MVVM perfectly separates data and logic from UI so it can be ported however you want in the future without having to rediscover and reimplement features. If you are new to MVVM read up on properties, bindings and commands (relay commands mostly). MVVM Toolkit is the easiest to use right now and I would recommend starting with that from the beginning.
 6. View code behind should ideally be non-existent (unlike WinForms). Everything is done in the view model. However there are always places where bindings and controls don't do exactly as you want. It is usually faster and easier to maintian to just handle things in event handlers at this point. However, behaviors is another layer you can pull in and use to keep 100% clean MVVM. In practice it's almost never worth it to write your own behaviors though when a simple event handler that everyone understands can get the job done.

As you go cross platform, remember Avalonia is a UI framework only and not a platform. So as you go cross platform you'll have to find abstraction libraries for various things not related to UI (or just write code yourself). .NET it getting better here slowly but integrating with various services cross platform is still tricky. I would also identify 3rd party libraries and features that are windows only. With old WinForms apps there might be a lot of Win32 code and libraries/controls you can't easily bring over.

**Reply by wsandy1** [reply]
Thanks again, all these tips are very helpful.
